{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "########################################################################\n",
    "# import default libraries\n",
    "########################################################################\n",
    "import os\n",
    "import sys\n",
    "import gc\n",
    "########################################################################\n",
    "\n",
    "\n",
    "########################################################################\n",
    "# import additional libraries\n",
    "########################################################################\n",
    "import numpy as np\n",
    "import scipy.stats\n",
    "# from import\n",
    "from tqdm import tqdm\n",
    "try:\n",
    "    from sklearn.externals import joblib\n",
    "except:\n",
    "    import joblib\n",
    "# original lib\n",
    "import common as com\n",
    "import keras_model\n",
    "import pandas as pd\n",
    "import keras\n",
    "import librosa\n",
    "import tensorflow as tf\n",
    "import librosa.display\n",
    "############\n",
    "########################################################################\n",
    "# load parameter.yaml\n",
    "########################################################################\n",
    "param = com.yaml_load()\n",
    "os.environ['TF_FORCE_GPU_ALLOW_GROWTH'] = 'true'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def spec_augment_freq(spec: np.ndarray, num_mask=1, freq_masking_max_percentage=0.9):\n",
    "    spec = spec.copy()\n",
    "    spec[24:33, :] = -0\n",
    "    spec[33:39, :] = -0\n",
    "    return spec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-10-08 16:46:08,435 - INFO - load_directory <- development\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/notebooks/dev_data/car1', '/notebooks/dev_data/car8']\n",
      "\n",
      "===========================\n",
      "[1/2] /notebooks/dev_data/car1\n",
      "============== MODEL LOAD ==============\n",
      "/notebooks/model/model_car1.hdf5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-10-08 16:46:09.635883: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:39] Overriding allow_growth setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model: \"model\"\n",
      "__________________________________________________________________________________________________\n",
      " Layer (type)                   Output Shape         Param #     Connected to                     \n",
      "==================================================================================================\n",
      " input_1 (InputLayer)           [(None, 256, 128, 1  0           []                               \n",
      "                                )]                                                                \n",
      "                                                                                                  \n",
      " block_1_conv (Conv2D)          (None, 128, 64, 16)  416         ['input_1[0][0]']                \n",
      "                                                                                                  \n",
      " block_1_conv_bn (BatchNormaliz  (None, 128, 64, 16)  64         ['block_1_conv[0][0]']           \n",
      " ation)                                                                                           \n",
      "                                                                                                  \n",
      " block_1_relu (ReLU)            (None, 128, 64, 16)  0           ['block_1_conv_bn[0][0]']        \n",
      "                                                                                                  \n",
      " block_2_conv (Conv2D)          (None, 64, 32, 32)   12832       ['block_1_relu[0][0]']           \n",
      "                                                                                                  \n",
      " block_2_conv_bn (BatchNormaliz  (None, 64, 32, 32)  128         ['block_2_conv[0][0]']           \n",
      " ation)                                                                                           \n",
      "                                                                                                  \n",
      " block_2_relu (ReLU)            (None, 64, 32, 32)   0           ['block_2_conv_bn[0][0]']        \n",
      "                                                                                                  \n",
      " block_3_conv (Conv2D)          (None, 32, 16, 64)   51264       ['block_2_relu[0][0]']           \n",
      "                                                                                                  \n",
      " block_3_conv_bn (BatchNormaliz  (None, 32, 16, 64)  256         ['block_3_conv[0][0]']           \n",
      " ation)                                                                                           \n",
      "                                                                                                  \n",
      " block_3_relu (ReLU)            (None, 32, 16, 64)   0           ['block_3_conv_bn[0][0]']        \n",
      "                                                                                                  \n",
      " block_4_conv (Conv2D)          (None, 16, 8, 128)   204928      ['block_3_relu[0][0]']           \n",
      "                                                                                                  \n",
      " block_4_conv_bn (BatchNormaliz  (None, 16, 8, 128)  512         ['block_4_conv[0][0]']           \n",
      " ation)                                                                                           \n",
      "                                                                                                  \n",
      " block_4_relu (ReLU)            (None, 16, 8, 128)   0           ['block_4_conv_bn[0][0]']        \n",
      "                                                                                                  \n",
      " block_6_conv (Conv2D)          (None, 8, 4, 512)    1638912     ['block_4_relu[0][0]']           \n",
      "                                                                                                  \n",
      " block_6_conv_bn (BatchNormaliz  (None, 8, 4, 512)   2048        ['block_6_conv[0][0]']           \n",
      " ation)                                                                                           \n",
      "                                                                                                  \n",
      " block_6_relu (ReLU)            (None, 8, 4, 512)    0           ['block_6_conv_bn[0][0]']        \n",
      "                                                                                                  \n",
      " block_7_de-conv (Conv2DTranspo  (None, 8, 4, 256)   3277056     ['block_6_relu[0][0]']           \n",
      " se)                                                                                              \n",
      "                                                                                                  \n",
      " block_5_conv (Conv2D)          (None, 8, 4, 256)    819456      ['block_4_relu[0][0]']           \n",
      "                                                                                                  \n",
      " block_7_conv_bn (BatchNormaliz  (None, 8, 4, 256)   1024        ['block_7_de-conv[0][0]']        \n",
      " ation)                                                                                           \n",
      "                                                                                                  \n",
      " block_5_conv_bn (BatchNormaliz  (None, 8, 4, 256)   1024        ['block_5_conv[0][0]']           \n",
      " ation)                                                                                           \n",
      "                                                                                                  \n",
      " block_7_lrelu (LeakyReLU)      (None, 8, 4, 256)    0           ['block_7_conv_bn[0][0]']        \n",
      "                                                                                                  \n",
      " block_5_relu (ReLU)            (None, 8, 4, 256)    0           ['block_5_conv_bn[0][0]']        \n",
      "                                                                                                  \n",
      " skip1 (Concatenate)            (None, 8, 4, 512)    0           ['block_7_lrelu[0][0]',          \n",
      "                                                                  'block_5_relu[0][0]']           \n",
      "                                                                                                  \n",
      " block_8_de-conv (Conv2DTranspo  (None, 16, 8, 128)  1638528     ['skip1[0][0]']                  \n",
      " se)                                                                                              \n",
      "                                                                                                  \n",
      " block_8_conv_bn (BatchNormaliz  (None, 16, 8, 128)  512         ['block_8_de-conv[0][0]']        \n",
      " ation)                                                                                           \n",
      "                                                                                                  \n",
      " block_8_lrelu (LeakyReLU)      (None, 16, 8, 128)   0           ['block_8_conv_bn[0][0]']        \n",
      "                                                                                                  \n",
      " skip2 (Concatenate)            (None, 16, 8, 256)   0           ['block_8_lrelu[0][0]',          \n",
      "                                                                  'block_4_relu[0][0]']           \n",
      "                                                                                                  \n",
      " block_9_de-conv (Conv2DTranspo  (None, 32, 16, 64)  409664      ['skip2[0][0]']                  \n",
      " se)                                                                                              \n",
      "                                                                                                  \n",
      " block_9_conv_bn (BatchNormaliz  (None, 32, 16, 64)  256         ['block_9_de-conv[0][0]']        \n",
      " ation)                                                                                           \n",
      "                                                                                                  \n",
      " block_9_lrelu (LeakyReLU)      (None, 32, 16, 64)   0           ['block_9_conv_bn[0][0]']        \n",
      "                                                                                                  \n",
      " skip3 (Concatenate)            (None, 32, 16, 128)  0           ['block_9_lrelu[0][0]',          \n",
      "                                                                  'block_3_relu[0][0]']           \n",
      "                                                                                                  \n",
      " block_10_de-conv (Conv2DTransp  (None, 64, 32, 32)  102432      ['skip3[0][0]']                  \n",
      " ose)                                                                                             \n",
      "                                                                                                  \n",
      " block_10_conv_bn (BatchNormali  (None, 64, 32, 32)  128         ['block_10_de-conv[0][0]']       \n",
      " zation)                                                                                          \n",
      "                                                                                                  \n",
      " block_10_lrelu (LeakyReLU)     (None, 64, 32, 32)   0           ['block_10_conv_bn[0][0]']       \n",
      "                                                                                                  \n",
      " concatenate (Concatenate)      (None, 64, 32, 64)   0           ['block_10_lrelu[0][0]',         \n",
      "                                                                  'block_2_relu[0][0]']           \n",
      "                                                                                                  \n",
      " block_12_de-conv (Conv2DTransp  (None, 256, 128, 1)  1601       ['concatenate[0][0]']            \n",
      " ose)                                                                                             \n",
      "                                                                                                  \n",
      " block_12_conv_bn (BatchNormali  (None, 256, 128, 1)  4          ['block_12_de-conv[0][0]']       \n",
      " zation)                                                                                          \n",
      "                                                                                                  \n",
      " block_12_lrelu (LeakyReLU)     (None, 256, 128, 1)  0           ['block_12_conv_bn[0][0]']       \n",
      "                                                                                                  \n",
      "==================================================================================================\n",
      "Total params: 8,163,045\n",
      "Trainable params: 8,160,067\n",
      "Non-trainable params: 2,978\n",
      "__________________________________________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "target_dir = \"/notebooks/dev_data/car1\"\n",
    "mode = True\n",
    "dirs = com.select_dirs(param=param, mode=mode)\n",
    "print(dirs)\n",
    "\n",
    "# loop of the base dire/notebooks/model/model_car1.hdf5ctory\n",
    "idx=0\n",
    "print(\"\\n===========================\")\n",
    "print(\"[{idx}/{total}] {target_dir}\".format(target_dir=target_dir, idx=idx+1, total=len(dirs)))\n",
    "machine_type = os.path.split(target_dir)[1]\n",
    "\n",
    "\n",
    "def load_model(file_path):\n",
    "    return keras.models.load_model(file_path, compile=False)\n",
    "\n",
    "  \n",
    "print(\"============== MODEL LOAD ==============\")\n",
    "# load model file\n",
    "model_file = \"{model}/model_{machine_type}.hdf5\".format(model=param[\"model_directory\"],\n",
    "                                                        machine_type=machine_type)\n",
    "print(model_file)\n",
    "if not os.path.exists(model_file):\n",
    "    com.logger.error(\"{} model not found \".format(machine_type))\n",
    "    sys.exit(-1)\n",
    "model = load_model(model_file)\n",
    "model.summary()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "============== BEGIN TEST FOR A SECTION ==============\n",
      "dev_data/car1/source_test/section_00_source_test_anomaly_0000 (107).wav\n",
      "(256, 128, 1)\n"
     ]
    }
   ],
   "source": [
    "print(\"\\n============== BEGIN TEST FOR A SECTION ==============\")\n",
    "y_pred = [0. for k in range(20)]\n",
    "y_pred_freq = [0. for k in range(20)]\n",
    "y_pred_time = [0. for k in range(20)]\n",
    "file_idx=1\n",
    "# file_path='/content/drive/MyDrive/dev_data/car8/source_test/section_07_source_test_anomaly_0000 (126).wav'\n",
    "file_path='dev_data/car1/source_test/section_00_source_test_anomaly_0000 (107).wav'\n",
    "print(file_path)\n",
    "# file_path='/content/drive/MyDrive/dev_data/car8/source_test/section_07_source_test_anomaly_0000 (7).wav'\n",
    "# file_path='D:/B4/u-net/dev_data/car1/source_test/section_00_source_test_normal_0000 (1).wav'\n",
    "vectors = com.file_to_vectors(file_path,\n",
    "                              n_mels=128,\n",
    "                              n_frames=1,\n",
    "                              n_fft=1024,\n",
    "                              hop_length=862,\n",
    "                              power=2)\n",
    "vectors = librosa.power_to_db(vectors)\n",
    "vectors_masked = spec_augment_freq(vectors)\n",
    "data = np.zeros((1, vectors.shape[1], vectors.shape[0]),float)\n",
    "data_masked = np.zeros((1, vectors_masked.shape[1], vectors_masked.shape[0]), float)\n",
    "data[0, :, :] = vectors.T\n",
    "data_masked[0, :, :] = vectors_masked.T\n",
    "data = data[:, :, :, np.newaxis]\n",
    "data_masked = data_masked[:, :, :, np.newaxis]\n",
    "\n",
    "\n",
    "\n",
    "print(data[0,:,:].shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-10-08 16:46:11.593319: E tensorflow/stream_executor/cuda/cuda_dnn.cc:373] Could not create cudnn handle: CUDNN_STATUS_INTERNAL_ERROR\n",
      "2022-10-08 16:46:11.593581: W tensorflow/core/framework/op_kernel.cc:1745] OP_REQUIRES failed at conv_ops.cc:1120 : UNIMPLEMENTED: DNN library is not found.\n"
     ]
    },
    {
     "ename": "UnimplementedError",
     "evalue": "Graph execution error:\n\nDetected at node 'model/block_1_conv/Conv2D' defined at (most recent call last):\n    File \"/usr/lib/python3.9/runpy.py\", line 197, in _run_module_as_main\n      return _run_code(code, main_globals, None,\n    File \"/usr/lib/python3.9/runpy.py\", line 87, in _run_code\n      exec(code, run_globals)\n    File \"/usr/local/lib/python3.9/dist-packages/ipykernel_launcher.py\", line 17, in <module>\n      app.launch_new_instance()\n    File \"/usr/local/lib/python3.9/dist-packages/traitlets/config/application.py\", line 976, in launch_instance\n      app.start()\n    File \"/usr/local/lib/python3.9/dist-packages/ipykernel/kernelapp.py\", line 712, in start\n      self.io_loop.start()\n    File \"/usr/local/lib/python3.9/dist-packages/tornado/platform/asyncio.py\", line 215, in start\n      self.asyncio_loop.run_forever()\n    File \"/usr/lib/python3.9/asyncio/base_events.py\", line 601, in run_forever\n      self._run_once()\n    File \"/usr/lib/python3.9/asyncio/base_events.py\", line 1905, in _run_once\n      handle._run()\n    File \"/usr/lib/python3.9/asyncio/events.py\", line 80, in _run\n      self._context.run(self._callback, *self._args)\n    File \"/usr/local/lib/python3.9/dist-packages/ipykernel/kernelbase.py\", line 510, in dispatch_queue\n      await self.process_one()\n    File \"/usr/local/lib/python3.9/dist-packages/ipykernel/kernelbase.py\", line 499, in process_one\n      await dispatch(*args)\n    File \"/usr/local/lib/python3.9/dist-packages/ipykernel/kernelbase.py\", line 406, in dispatch_shell\n      await result\n    File \"/usr/local/lib/python3.9/dist-packages/ipykernel/kernelbase.py\", line 730, in execute_request\n      reply_content = await reply_content\n    File \"/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py\", line 383, in do_execute\n      res = shell.run_cell(\n    File \"/usr/local/lib/python3.9/dist-packages/ipykernel/zmqshell.py\", line 528, in run_cell\n      return super().run_cell(*args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/IPython/core/interactiveshell.py\", line 2881, in run_cell\n      result = self._run_cell(\n    File \"/usr/local/lib/python3.9/dist-packages/IPython/core/interactiveshell.py\", line 2936, in _run_cell\n      return runner(coro)\n    File \"/usr/local/lib/python3.9/dist-packages/IPython/core/async_helpers.py\", line 129, in _pseudo_sync_runner\n      coro.send(None)\n    File \"/usr/local/lib/python3.9/dist-packages/IPython/core/interactiveshell.py\", line 3135, in run_cell_async\n      has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n    File \"/usr/local/lib/python3.9/dist-packages/IPython/core/interactiveshell.py\", line 3338, in run_ast_nodes\n      if await self.run_code(code, result, async_=asy):\n    File \"/usr/local/lib/python3.9/dist-packages/IPython/core/interactiveshell.py\", line 3398, in run_code\n      exec(code_obj, self.user_global_ns, self.user_ns)\n    File \"/tmp/ipykernel_1634/379540993.py\", line 1, in <cell line: 1>\n      reconstruction = model.predict(data_masked)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/utils/traceback_utils.py\", line 64, in error_handler\n      return fn(*args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/training.py\", line 1982, in predict\n      tmp_batch_outputs = self.predict_function(iterator)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/training.py\", line 1801, in predict_function\n      return step_function(self, iterator)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/training.py\", line 1790, in step_function\n      outputs = model.distribute_strategy.run(run_step, args=(data,))\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/training.py\", line 1783, in run_step\n      outputs = model.predict_step(data)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/training.py\", line 1751, in predict_step\n      return self(x, training=False)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/utils/traceback_utils.py\", line 64, in error_handler\n      return fn(*args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/base_layer.py\", line 1096, in __call__\n      outputs = call_fn(inputs, *args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/utils/traceback_utils.py\", line 92, in error_handler\n      return fn(*args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/functional.py\", line 451, in call\n      return self._run_internal_graph(\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/functional.py\", line 589, in _run_internal_graph\n      outputs = node.layer(*args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/utils/traceback_utils.py\", line 64, in error_handler\n      return fn(*args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/base_layer.py\", line 1096, in __call__\n      outputs = call_fn(inputs, *args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/utils/traceback_utils.py\", line 92, in error_handler\n      return fn(*args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/layers/convolutional.py\", line 248, in call\n      outputs = self.convolution_op(inputs, self.kernel)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/layers/convolutional.py\", line 233, in convolution_op\n      return tf.nn.convolution(\nNode: 'model/block_1_conv/Conv2D'\nDNN library is not found.\n\t [[{{node model/block_1_conv/Conv2D}}]] [Op:__inference_predict_function_1209]",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mUnimplementedError\u001b[0m                        Traceback (most recent call last)",
      "Input \u001b[0;32mIn [6]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m reconstruction \u001b[38;5;241m=\u001b[39m \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpredict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata_masked\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m y_pred \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mmean(np\u001b[38;5;241m.\u001b[39msquare((data[\u001b[38;5;241m0\u001b[39m,:,\u001b[38;5;241m24\u001b[39m:\u001b[38;5;241m39\u001b[39m] \u001b[38;5;241m-\u001b[39m reconstruction[\u001b[38;5;241m0\u001b[39m,:,\u001b[38;5;241m24\u001b[39m:\u001b[38;5;241m39\u001b[39m])))\n\u001b[1;32m      3\u001b[0m plt\u001b[38;5;241m.\u001b[39mgcf()\u001b[38;5;241m.\u001b[39mset_size_inches(\u001b[38;5;241m8\u001b[39m, \u001b[38;5;241m8\u001b[39m)\n",
      "File \u001b[0;32m/usr/local/lib/python3.9/dist-packages/keras/utils/traceback_utils.py:67\u001b[0m, in \u001b[0;36mfilter_traceback.<locals>.error_handler\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     65\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:  \u001b[38;5;66;03m# pylint: disable=broad-except\u001b[39;00m\n\u001b[1;32m     66\u001b[0m   filtered_tb \u001b[38;5;241m=\u001b[39m _process_traceback_frames(e\u001b[38;5;241m.\u001b[39m__traceback__)\n\u001b[0;32m---> 67\u001b[0m   \u001b[38;5;28;01mraise\u001b[39;00m e\u001b[38;5;241m.\u001b[39mwith_traceback(filtered_tb) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28mNone\u001b[39m\n\u001b[1;32m     68\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m     69\u001b[0m   \u001b[38;5;28;01mdel\u001b[39;00m filtered_tb\n",
      "File \u001b[0;32m/usr/local/lib/python3.9/dist-packages/tensorflow/python/eager/execute.py:54\u001b[0m, in \u001b[0;36mquick_execute\u001b[0;34m(op_name, num_outputs, inputs, attrs, ctx, name)\u001b[0m\n\u001b[1;32m     52\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m     53\u001b[0m   ctx\u001b[38;5;241m.\u001b[39mensure_initialized()\n\u001b[0;32m---> 54\u001b[0m   tensors \u001b[38;5;241m=\u001b[39m pywrap_tfe\u001b[38;5;241m.\u001b[39mTFE_Py_Execute(ctx\u001b[38;5;241m.\u001b[39m_handle, device_name, op_name,\n\u001b[1;32m     55\u001b[0m                                       inputs, attrs, num_outputs)\n\u001b[1;32m     56\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m core\u001b[38;5;241m.\u001b[39m_NotOkStatusException \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m     57\u001b[0m   \u001b[38;5;28;01mif\u001b[39;00m name \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "\u001b[0;31mUnimplementedError\u001b[0m: Graph execution error:\n\nDetected at node 'model/block_1_conv/Conv2D' defined at (most recent call last):\n    File \"/usr/lib/python3.9/runpy.py\", line 197, in _run_module_as_main\n      return _run_code(code, main_globals, None,\n    File \"/usr/lib/python3.9/runpy.py\", line 87, in _run_code\n      exec(code, run_globals)\n    File \"/usr/local/lib/python3.9/dist-packages/ipykernel_launcher.py\", line 17, in <module>\n      app.launch_new_instance()\n    File \"/usr/local/lib/python3.9/dist-packages/traitlets/config/application.py\", line 976, in launch_instance\n      app.start()\n    File \"/usr/local/lib/python3.9/dist-packages/ipykernel/kernelapp.py\", line 712, in start\n      self.io_loop.start()\n    File \"/usr/local/lib/python3.9/dist-packages/tornado/platform/asyncio.py\", line 215, in start\n      self.asyncio_loop.run_forever()\n    File \"/usr/lib/python3.9/asyncio/base_events.py\", line 601, in run_forever\n      self._run_once()\n    File \"/usr/lib/python3.9/asyncio/base_events.py\", line 1905, in _run_once\n      handle._run()\n    File \"/usr/lib/python3.9/asyncio/events.py\", line 80, in _run\n      self._context.run(self._callback, *self._args)\n    File \"/usr/local/lib/python3.9/dist-packages/ipykernel/kernelbase.py\", line 510, in dispatch_queue\n      await self.process_one()\n    File \"/usr/local/lib/python3.9/dist-packages/ipykernel/kernelbase.py\", line 499, in process_one\n      await dispatch(*args)\n    File \"/usr/local/lib/python3.9/dist-packages/ipykernel/kernelbase.py\", line 406, in dispatch_shell\n      await result\n    File \"/usr/local/lib/python3.9/dist-packages/ipykernel/kernelbase.py\", line 730, in execute_request\n      reply_content = await reply_content\n    File \"/usr/local/lib/python3.9/dist-packages/ipykernel/ipkernel.py\", line 383, in do_execute\n      res = shell.run_cell(\n    File \"/usr/local/lib/python3.9/dist-packages/ipykernel/zmqshell.py\", line 528, in run_cell\n      return super().run_cell(*args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/IPython/core/interactiveshell.py\", line 2881, in run_cell\n      result = self._run_cell(\n    File \"/usr/local/lib/python3.9/dist-packages/IPython/core/interactiveshell.py\", line 2936, in _run_cell\n      return runner(coro)\n    File \"/usr/local/lib/python3.9/dist-packages/IPython/core/async_helpers.py\", line 129, in _pseudo_sync_runner\n      coro.send(None)\n    File \"/usr/local/lib/python3.9/dist-packages/IPython/core/interactiveshell.py\", line 3135, in run_cell_async\n      has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n    File \"/usr/local/lib/python3.9/dist-packages/IPython/core/interactiveshell.py\", line 3338, in run_ast_nodes\n      if await self.run_code(code, result, async_=asy):\n    File \"/usr/local/lib/python3.9/dist-packages/IPython/core/interactiveshell.py\", line 3398, in run_code\n      exec(code_obj, self.user_global_ns, self.user_ns)\n    File \"/tmp/ipykernel_1634/379540993.py\", line 1, in <cell line: 1>\n      reconstruction = model.predict(data_masked)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/utils/traceback_utils.py\", line 64, in error_handler\n      return fn(*args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/training.py\", line 1982, in predict\n      tmp_batch_outputs = self.predict_function(iterator)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/training.py\", line 1801, in predict_function\n      return step_function(self, iterator)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/training.py\", line 1790, in step_function\n      outputs = model.distribute_strategy.run(run_step, args=(data,))\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/training.py\", line 1783, in run_step\n      outputs = model.predict_step(data)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/training.py\", line 1751, in predict_step\n      return self(x, training=False)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/utils/traceback_utils.py\", line 64, in error_handler\n      return fn(*args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/base_layer.py\", line 1096, in __call__\n      outputs = call_fn(inputs, *args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/utils/traceback_utils.py\", line 92, in error_handler\n      return fn(*args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/functional.py\", line 451, in call\n      return self._run_internal_graph(\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/functional.py\", line 589, in _run_internal_graph\n      outputs = node.layer(*args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/utils/traceback_utils.py\", line 64, in error_handler\n      return fn(*args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/engine/base_layer.py\", line 1096, in __call__\n      outputs = call_fn(inputs, *args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/utils/traceback_utils.py\", line 92, in error_handler\n      return fn(*args, **kwargs)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/layers/convolutional.py\", line 248, in call\n      outputs = self.convolution_op(inputs, self.kernel)\n    File \"/usr/local/lib/python3.9/dist-packages/keras/layers/convolutional.py\", line 233, in convolution_op\n      return tf.nn.convolution(\nNode: 'model/block_1_conv/Conv2D'\nDNN library is not found.\n\t [[{{node model/block_1_conv/Conv2D}}]] [Op:__inference_predict_function_1209]"
     ]
    }
   ],
   "source": [
    "reconstruction = model.predict(data_masked)\n",
    "y_pred = np.mean(np.square((data[0,:,24:39] - reconstruction[0,:,24:39])))\n",
    "plt.gcf().set_size_inches(8, 8)\n",
    "plt.subplot()\n",
    "librosa.display.specshow(np.squeeze(data, axis=3)[0,:,:].T, cmap='magma')\n",
    "plt.show()\n",
    "\n",
    "plt.subplot()\n",
    "plt.gcf().set_size_inches(8, 8)\n",
    "\n",
    "librosa.display.specshow(np.squeeze(data_masked, axis=3)[0,:,:].T, cmap='magma')\n",
    "plt.ylabel(\"Original\")\n",
    "plt.show()\n",
    "plt.subplot()\n",
    "plt.gcf().set_size_inches(8, 8)\n",
    "\n",
    "librosa.display.specshow(np.squeeze(reconstruction, axis=3)[0,:,:].T[:128,:], cmap='magma')\n",
    "plt.ylabel(\"Reconstruction\")\n",
    "# plt.subplot(3,1,3)\n",
    "# y_ano = []\n",
    "# y_nor = []\n",
    "# x_ano = []\n",
    "# x_nor = []\n",
    "# n=-1\n",
    "# for k in y_pred_time:\n",
    "#     n+=1\n",
    "#     if k > 13.3:\n",
    "#         x_ano.append(n)\n",
    "#         y_ano.append(1)\n",
    "# n=-1\n",
    "# for k in y_pred_time:\n",
    "#     n+=1\n",
    "#     if k < 13.3:\n",
    "#         x_nor.append(n)\n",
    "#         y_nor.append(0)\n",
    "# x_ano = [x*5*3*512/16000 for x in x_ano]\n",
    "# x_nor = [x*5*3*512/16000 for x in x_nor]\n",
    "# x_nor[-1] = 10\n",
    "# x_nor[0] = x_ano[-1]\n",
    "# plt.plot(x_ano, y_ano, '-',color='r')\n",
    "# plt.plot(x_nor, y_nor, '-',color='b')\n",
    "# plt.fill_between(x_ano, y_ano,-2, alpha= 0.2, label='Anomaly', color='r')\n",
    "# plt.ylim(0, 2)\n",
    "# plt.xlim(0, 10)\n",
    "# plt.xticks([])\n",
    "# plt.yticks([])\n",
    "# plt.legend(fontsize=8.5)\n",
    "# plt.subplot(4,1,4)\n",
    "# plt.plot(x, y_pred_time)\n",
    "# plt.plot(x, y, '--',color='r')\n",
    "# plt.ylim(0, max(y_pred_time)+10)\n",
    "# plt.xlim(0, 10)\n",
    "# plt.xlabel(\"Time (s)\")\n",
    "# plt.ylabel(\"Anomaly Score\")\n",
    "# plt.subplots_adjust(hspace=.0)\n",
    "# plt.fill_between(x, y_pred_time, y, alpha= 0.2, where = (np.array(y_pred_time) > 12.5), label='Anomaly', color='r')\n",
    "# plt.plot([], [], ' ', label=\"Mean Anomaly Score={:.2f}\".format(anomaly_score_list[1][1]))\n",
    "# plt.legend(fontsize=6.5)\n",
    "plt.show()\n",
    "print(y_pred)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting tensorflow==2.8\n",
      "  Downloading tensorflow-2.8.0-cp39-cp39-manylinux2010_x86_64.whl (497.6 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m497.6/497.6 MB\u001b[0m \u001b[31m3.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: grpcio<2.0,>=1.24.3 in /usr/local/lib/python3.9/dist-packages (from tensorflow==2.8) (1.47.0)\n",
      "Requirement already satisfied: flatbuffers>=1.12 in /usr/local/lib/python3.9/dist-packages (from tensorflow==2.8) (1.12)\n",
      "Requirement already satisfied: libclang>=9.0.1 in /usr/local/lib/python3.9/dist-packages (from tensorflow==2.8) (14.0.1)\n",
      "Requirement already satisfied: setuptools in /usr/local/lib/python3.9/dist-packages (from tensorflow==2.8) (59.8.0)\n",
      "Requirement already satisfied: astunparse>=1.6.0 in /usr/local/lib/python3.9/dist-packages (from tensorflow==2.8) (1.6.3)\n",
      "Requirement already satisfied: tensorflow-io-gcs-filesystem>=0.23.1 in /usr/local/lib/python3.9/dist-packages (from tensorflow==2.8) (0.26.0)\n",
      "Requirement already satisfied: typing-extensions>=3.6.6 in /usr/local/lib/python3.9/dist-packages (from tensorflow==2.8) (4.3.0)\n",
      "Requirement already satisfied: protobuf>=3.9.2 in /usr/local/lib/python3.9/dist-packages (from tensorflow==2.8) (3.19.4)\n",
      "Collecting tensorboard<2.9,>=2.8\n",
      "  Downloading tensorboard-2.8.0-py3-none-any.whl (5.8 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m5.8/5.8 MB\u001b[0m \u001b[31m62.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: absl-py>=0.4.0 in /usr/local/lib/python3.9/dist-packages (from tensorflow==2.8) (1.1.0)\n",
      "Requirement already satisfied: h5py>=2.9.0 in /usr/local/lib/python3.9/dist-packages (from tensorflow==2.8) (3.7.0)\n",
      "Requirement already satisfied: termcolor>=1.1.0 in /usr/local/lib/python3.9/dist-packages (from tensorflow==2.8) (1.1.0)\n",
      "Collecting tf-estimator-nightly==2.8.0.dev2021122109\n",
      "  Downloading tf_estimator_nightly-2.8.0.dev2021122109-py2.py3-none-any.whl (462 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m462.5/462.5 kB\u001b[0m \u001b[31m53.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: numpy>=1.20 in /usr/local/lib/python3.9/dist-packages (from tensorflow==2.8) (1.23.1)\n",
      "Requirement already satisfied: keras-preprocessing>=1.1.1 in /usr/local/lib/python3.9/dist-packages (from tensorflow==2.8) (1.1.2)\n",
      "Requirement already satisfied: google-pasta>=0.1.1 in /usr/local/lib/python3.9/dist-packages (from tensorflow==2.8) (0.2.0)\n",
      "Requirement already satisfied: six>=1.12.0 in /usr/lib/python3/dist-packages (from tensorflow==2.8) (1.14.0)\n",
      "Collecting keras<2.9,>=2.8.0rc0\n",
      "  Downloading keras-2.8.0-py2.py3-none-any.whl (1.4 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.4/1.4 MB\u001b[0m \u001b[31m64.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: opt-einsum>=2.3.2 in /usr/local/lib/python3.9/dist-packages (from tensorflow==2.8) (3.3.0)\n",
      "Requirement already satisfied: gast>=0.2.1 in /usr/local/lib/python3.9/dist-packages (from tensorflow==2.8) (0.4.0)\n",
      "Requirement already satisfied: wrapt>=1.11.0 in /usr/local/lib/python3.9/dist-packages (from tensorflow==2.8) (1.14.1)\n",
      "Requirement already satisfied: wheel<1.0,>=0.23.0 in /usr/local/lib/python3.9/dist-packages (from astunparse>=1.6.0->tensorflow==2.8) (0.35.1)\n",
      "Requirement already satisfied: markdown>=2.6.8 in /usr/local/lib/python3.9/dist-packages (from tensorboard<2.9,>=2.8->tensorflow==2.8) (3.3.7)\n",
      "Requirement already satisfied: requests<3,>=2.21.0 in /usr/local/lib/python3.9/dist-packages (from tensorboard<2.9,>=2.8->tensorflow==2.8) (2.28.1)\n",
      "Requirement already satisfied: tensorboard-plugin-wit>=1.6.0 in /usr/local/lib/python3.9/dist-packages (from tensorboard<2.9,>=2.8->tensorflow==2.8) (1.8.1)\n",
      "Requirement already satisfied: google-auth<3,>=1.6.3 in /usr/local/lib/python3.9/dist-packages (from tensorboard<2.9,>=2.8->tensorflow==2.8) (2.9.0)\n",
      "Requirement already satisfied: tensorboard-data-server<0.7.0,>=0.6.0 in /usr/local/lib/python3.9/dist-packages (from tensorboard<2.9,>=2.8->tensorflow==2.8) (0.6.1)\n",
      "Requirement already satisfied: werkzeug>=0.11.15 in /usr/local/lib/python3.9/dist-packages (from tensorboard<2.9,>=2.8->tensorflow==2.8) (2.1.2)\n",
      "Requirement already satisfied: google-auth-oauthlib<0.5,>=0.4.1 in /usr/local/lib/python3.9/dist-packages (from tensorboard<2.9,>=2.8->tensorflow==2.8) (0.4.6)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in /usr/lib/python3/dist-packages (from google-auth<3,>=1.6.3->tensorboard<2.9,>=2.8->tensorflow==2.8) (4.0)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /usr/local/lib/python3.9/dist-packages (from google-auth<3,>=1.6.3->tensorboard<2.9,>=2.8->tensorflow==2.8) (0.2.8)\n",
      "Requirement already satisfied: cachetools<6.0,>=2.0.0 in /usr/local/lib/python3.9/dist-packages (from google-auth<3,>=1.6.3->tensorboard<2.9,>=2.8->tensorflow==2.8) (5.2.0)\n",
      "Requirement already satisfied: requests-oauthlib>=0.7.0 in /usr/local/lib/python3.9/dist-packages (from google-auth-oauthlib<0.5,>=0.4.1->tensorboard<2.9,>=2.8->tensorflow==2.8) (1.3.1)\n",
      "Requirement already satisfied: importlib-metadata>=4.4 in /usr/local/lib/python3.9/dist-packages (from markdown>=2.6.8->tensorboard<2.9,>=2.8->tensorflow==2.8) (4.12.0)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /usr/lib/python3/dist-packages (from requests<3,>=2.21.0->tensorboard<2.9,>=2.8->tensorflow==2.8) (2.8)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /usr/lib/python3/dist-packages (from requests<3,>=2.21.0->tensorboard<2.9,>=2.8->tensorflow==2.8) (2019.11.28)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /usr/local/lib/python3.9/dist-packages (from requests<3,>=2.21.0->tensorboard<2.9,>=2.8->tensorflow==2.8) (1.26.10)\n",
      "Requirement already satisfied: charset-normalizer<3,>=2 in /usr/local/lib/python3.9/dist-packages (from requests<3,>=2.21.0->tensorboard<2.9,>=2.8->tensorflow==2.8) (2.1.0)\n",
      "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.9/dist-packages (from importlib-metadata>=4.4->markdown>=2.6.8->tensorboard<2.9,>=2.8->tensorflow==2.8) (3.8.1)\n",
      "Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /usr/local/lib/python3.9/dist-packages (from pyasn1-modules>=0.2.1->google-auth<3,>=1.6.3->tensorboard<2.9,>=2.8->tensorflow==2.8) (0.4.8)\n",
      "Requirement already satisfied: oauthlib>=3.0.0 in /usr/local/lib/python3.9/dist-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<0.5,>=0.4.1->tensorboard<2.9,>=2.8->tensorflow==2.8) (3.2.0)\n",
      "Installing collected packages: tf-estimator-nightly, keras, tensorboard, tensorflow\n",
      "  Attempting uninstall: keras\n",
      "    Found existing installation: keras 2.7.0\n",
      "    Uninstalling keras-2.7.0:\n",
      "      Successfully uninstalled keras-2.7.0\n",
      "  Attempting uninstall: tensorboard\n",
      "    Found existing installation: tensorboard 2.9.1\n",
      "    Uninstalling tensorboard-2.9.1:\n",
      "      Successfully uninstalled tensorboard-2.9.1\n",
      "  Attempting uninstall: tensorflow\n",
      "    Found existing installation: tensorflow 2.7.0\n",
      "    Uninstalling tensorflow-2.7.0:\n",
      "      Successfully uninstalled tensorflow-2.7.0\n",
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "tensorflow-gpu 2.7.0 requires keras<2.8,>=2.7.0rc0, but you have keras 2.8.0 which is incompatible.\u001b[0m\u001b[31m\n",
      "\u001b[0mSuccessfully installed keras-2.8.0 tensorboard-2.8.0 tensorflow-2.8.0 tf-estimator-nightly-2.8.0.dev2021122109\n",
      "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\u001b[33m\n",
      "Reading package lists... Done\n",
      "Building dependency tree       \n",
      "Reading state information... Done\n",
      "\u001b[1;31mE: \u001b[0mVersion '8.1.0.77-1+cuda11.2' for 'libcudnn8' was not found\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install tensorflow==2.8\n",
    "!apt install --allow-change-held-packages libcudnn8=8.1.0.77-1+cuda11.2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "vscode": {
   "interpreter": {
    "hash": "a7d30985f23d0c7700ff9117488976e591efd57aed74d4a8ca9ee5ad3f3e7e0d"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
